pkg/serving/config_changes.go

```diff
@@ -152,6 +152,22 @@ func UpdateResources(template *servingv1alpha1.RevisionTemplateSpec, requestsRes
 	return nil
 }
 
+// UpdateServiceLabels updates the labels on a service
+func UpdateServiceLabels(service *servingv1alpha1.Service, vars map[string]string) error {
+	if service.ObjectMeta.Labels == nil {
+		service.ObjectMeta.Labels = make(map[string]string)
```

Comment:
The question is whether the labels of the service or the labels of the revision template which should be set.

Reply:
I think this should probably be decided in #328. I like the simplicity of `--label` setting both, but not sure if preventing users from setting separate labels on services and revision templates would be an issue.

Reply:
> Do you have any suggestions on an alternative way to unset env vars and labels? I suppose a simple option is to add a `--remove-[env|label]` flag. It looks like [kubectl has you to pass `VAR-`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#-em-env-em-) where the minus sign indicates removal.

yes, I think we should settle on a similar convention. It's just people sometimes use just empty env vars as kind of a boolean flag (not recommended though). But we need a way to differentiate between an empty value (key with empty value) and not-existing entry (no key)

Reply:
My gut feeling is that people most often want to set labels for their pods, not the service. So we could settle on something like that:

* `--label mylabel=myvalue` would set it on both, the service and the pod
* `--label ^mylabel=myvalue` would set it only on the `Service` (the "outer" entity)
* `--label  :mylabel=myvalue` would set it only on the revision template (the "inner" entity)

That would cover 90% of all use case by setting both, but allows for tuning, too. Not sure about te concrete characters (one should also check whether there are special shell meanings so that those chars can be used unescaped), but using any character which can not be used within label keys should be cool.

wdyt ?

Reply:
It seems at that point there might as well be separate `--revision-label`, `--service-label` , and `--label` flags. It would be much more explicit both in documentation and I think code that way. Plus, special characters change (slightly) across different shell environments and if [this list](https://docstore.mik.ua/orelly/unix/upt/ch08_19.htm) is any indication, it may not be worth the time and effort to find and confirm which characters are valid in a reasonable number of shell environments.

Reply:
I too like the simplicity of `—label` for all labels as the object is already decided in the command group. Adding more complexity and notations to remember is unnecessary as it complicates the “language” of the CLI IMO.

Reply:
> It seems at that point there might as well be separate `--revision-label`, `--service-label` , and `--label` flags. It would be much more explicit both in documentation and I think code that way. Plus, special characters change (slightly) across different shell environments and if [this list](https://docstore.mik.ua/orelly/unix/upt/ch08_19.htm) is any indication, it may not be worth the time and effort to find and confirm which characters are valid in a reasonable number of shell environments.

This is always a tradeoff between options bloat and being to obfuscated. Currently there is a tendency to add 1-5 new options per new features, and I'm doing my best to fight against the bloat. At least asking, whether there could be a compromise and whether all those options are needed (e.g. see https://github.com/knative/client/pull/345#issuecomment-519385669 or https://github.com/knative/client/pull/326#pullrequestreview-270639537 or https://github.com/knative/client/pull/282#pullrequestreview-271866769 how to reduce the UX surface and so the cognitive load). 

Of course, we should no go over the top, and using special chars like that might be the case here (although `^` and `:` are safe shell chars so no need to escape).

What's about that compromise: Let's implement `--label` to label both, service and template and wait on feedback if the other use cases (individual labelling of serving/template) are actually requested ?

Reply:
From a single feature's PR POV adding 3 new options might not be a big deal, but its the sum which might lead to multi page help outputs which are really overwhelming (especially when alphabetically ordered, and `--revision-label`, `--sevice-label` and `--label` will apear on totally different positions in a list of 50+ options)

Reply:
OK on compromise!

Reply:
Aside: I'm starting to think we should implement our own flags groupings, which might help with this.

Reply:
spf13/cobra#836 has some discussion on flag groupings in help text. Unfortunately, it doesn’t look like there’s currently first class support for it.